quantum/wpm.c

```diff
@@ -68,33 +92,62 @@ __attribute__((weak)) uint8_t wpm_regress_count(uint16_t keycode) {
 }
 #endif
 
+// Outside 'raw' mode we smooth results over time.
+
 void update_wpm(uint16_t keycode) {
     if (wpm_keycode(keycode)) {
-        if (wpm_timer > 0) {
-            uint16_t latest_wpm = 60000 / timer_elapsed(wpm_timer) / WPM_ESTIMATED_WORD_SIZE;
-            if (latest_wpm > UINT8_MAX) {
-                latest_wpm = UINT8_MAX;
-            }
-            current_wpm += ceilf((latest_wpm - current_wpm) * wpm_smoothing);
-        }
-        wpm_timer = timer_read();
+        period_presses[current_period]++;
     }
 #ifdef WPM_ALLOW_COUNT_REGRESSION
     uint8_t regress = wpm_regress_count(keycode);
     if (regress) {
-        if (current_wpm < regress) {
-            current_wpm = 0;
-        } else {
-            current_wpm -= regress;
-        }
-        wpm_timer = timer_read();
+        period_presses[current_period]--;
     }
 #endif
 }
 
 void decay_wpm(void) {
-    if (timer_elapsed(wpm_timer) > 1000) {
-        current_wpm += (-current_wpm) * wpm_smoothing;
-        wpm_timer = timer_read();
+    uint32_t presses = period_presses[0];
+    for (int i = 1; i <= periods; i++) {
+        presses += period_presses[i];
     }
```

Comment:
```suggestion
    int32_t presses = period_presses[0];
    for (int i = 1; i <= periods; i++) {
        presses += period_presses[i];
    }
    if (presses < 0) {
        presses = 0;
    }
```
With `WPM_ALLOW_COUNT_REGRESSION` and the changed `update_wpm()` code the `presses` count may end up negative.
(Not actually tested.)

Reply:
Related to this comment, I'm noticing that the current version of this commit is assuming that the user won't ever have more than 127 keypresses during a sampling period;  it overflows to negative if that happens.

By default, sampling periods are 0.1 seconds long (50 sampling periods covering 5 seconds), and so you'd need to be averaging around 15,240 wpm  to overflow a sampling period.

(maths:  127 keypresses maximum per 0.1 second equals 1,270 per second equals 76,200 per minute equals 15,240 words per minute at 5 keys per word)

That's a couple orders of magnitude higher than anyone should actually be able to type and so is probably completely safe with default configuration, but that sampling period and number of sampling buffers can be configured per keyboard, which could theoretically make the overflow point a whole lot easier to hit, especially if somebody is scraping for space in their firmware and reducing the number of wpm sampling buffers.  

How much do folks care about this?  Should I be explicitly guarding against overflow in the code (at the expense of a little more code to clamp values)?  Or alternately I could move to larger storage for each sample period and fewer periods by default?  I mean, you're never going to have someone making 32k keypresses in 0.2 seconds, if I was to double the storage-per-sampling-period and halve the number of sampling periods.  It'd probably be a lot less fragile to extreme config adjustments just in general.

Definitely interested in people's thoughts on this, though!

(the change request above re: handling negative values in the sampling periods is also a very valid issue and definitely needs fixing as well!  I just wanted to add this comment to the pile as well.)

Reply:
Especially since it is configurable, would be ideal to handle that overflow.